File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.java

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());

        for (int t = 0; t < T; t++) {
            String words = br.readLine();
            int K = Integer.parseInt(br.readLine());

            int mn = 10001;
            int mx = 0;

            List<Integer>[] idxs = new ArrayList[26];
            for (int i = 0; i < 26; i ++ ) {
                idxs[i] = new ArrayList<>();
            }
            for (int i = 0; i < words.length(); i++) {
                int cur = (int) words.charAt(i) - 97;
                idxs[cur].add(i);
            }

            for (List<Integer> idx : idxs) {
                for (int i = 0; i < idx.size() - K + 1; i++ ) {
                    int len = idx.get(i + K - 1) - idx.get(i) +1;
                    mn = Math.min(mn,len);
                    mx = Math.max(mx,len);
                }
            }

            if (mn == 10001 && mx == 0) {
                System.out.println(-1);
            } else {
                System.out.println(mn + " " + mx);
            }
        }
    }

}

File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.py

T = int(input())
for _ in range(T) :
    words = input()
    K = int(input())

    idxs = [[] for _ in range(26)]
    mn = 100001
    mx = 0

    for i, word in enumerate(words) :
        cur = ord(word) - 97
        idxs[cur].append(i)

    for idx in idxs :
        for i in range(len(idx)-K+1) :
            length = idx[i+K-1] - idx[i] + 1
            mn = min(mn,length)
            mx = max(mx,length)

    if mn == 100001 and mx == 0 :
        print(-1)
    else :
        print(mn, mx)
